cob_twist_controller/src/debug/debug_fk_vel_recursive_node.cpp

```diff
@@ -0,0 +1,258 @@
+        KDL::JntArray q_actual_dot = KDL::JntArray(chain_.getNrOfJoints());
+
+        for (unsigned int i = 0; i < chain_.getNrOfSegments(); i++)
+        {
+            KDL::Joint joint = chain_.getSegment(i).getJoint();
+            std::string joint_name = joint.getName();
+            if (joint.getType() == KDL::Joint::None)
+            {
+                ROS_DEBUG_STREAM("Skip fixed Joint '" << joint_name);
+                continue;
+            }
+            else
+            {
+                if (std::find(msg->joint_names.begin(), msg->joint_names.end(), joint_name) != msg->joint_names.end())
+                {
+                    unsigned int index = std::distance(msg->joint_names.begin(), std::find(msg->joint_names.begin(), msg->joint_names.end(), joint_name));
+                    q_desired(index) = msg->desired.positions[index];
+                    q_desired_dot(index) = msg->desired.velocities[index];
+                    q_actual(index) = msg->actual.positions[index];
+                    q_actual_dot(index) = msg->actual.velocities[index];
+                }
+                else
+                {
+                    ROS_ERROR_STREAM("Joint '" << joint_name << "' not found in JointTrajectoryControllerState");
+                    return;
+                }
+            }
+        }
+
+        /// compute desired twists recursively
+        std::vector< KDL::FrameVel > v_FrameVel_desired;
+        KDL::JntArrayVel jntArrayVel_desired = KDL::JntArrayVel(q_desired, q_desired_dot);
+        v_FrameVel_desired.resize(chain_.getNrOfSegments());
+        if (p_fksolver_vel_->JntToCart(jntArrayVel_desired, v_FrameVel_desired, chain_.getNrOfSegments()) >= 0)
+        {
+            sensor_msgs::JointState magnitude_desired_msg;
+            magnitude_desired_msg.header.stamp = msg->header.stamp;
+            for (unsigned int i = 0; i < chain_.getNrOfSegments(); i++)
+            {
+                magnitude_desired_msg.name.push_back(chain_.getSegment(i).getName());
+                magnitude_desired_msg.velocity.push_back(v_FrameVel_desired[i].GetTwist().vel.Norm());
+            }
+            twist_magnitude_desired_pub_.publish(magnitude_desired_msg);
+        }
+        /// compute actual twists recursively
+        std::vector< KDL::FrameVel > v_FrameVel_actual;
+        KDL::JntArrayVel jntArrayVel_actual = KDL::JntArrayVel(q_actual, q_actual_dot);
+        v_FrameVel_actual.resize(chain_.getNrOfSegments());
+        if (p_fksolver_vel_->JntToCart(jntArrayVel_actual, v_FrameVel_actual, chain_.getNrOfSegments()) >= 0)
+        {
+            sensor_msgs::JointState magnitude_actual_msg;
+            magnitude_actual_msg.header.stamp = msg->header.stamp;
+            for (unsigned int i = 0; i < chain_.getNrOfSegments(); i++)
+            {
+                magnitude_actual_msg.name.push_back(chain_.getSegment(i).getName());
+                magnitude_actual_msg.velocity.push_back(v_FrameVel_actual[i].GetTwist().vel.Norm());
```

Comment:
Not sure if taking the Norm is valid when we want to use derive the acceleration from these later on. 
Why take the ||norm||? Having a negative velocity should be fine, right?

Reply:
derivatives are not of interest for the time being....magnitudes are fine for the safety assessment

Reply:
Allright then. But still to keep in mind maybe.